core/src/main/java/org/apache/iceberg/DefaultValueParser.java

```diff
@@ -0,0 +1,363 @@
+          Types.NestedField field = fields.get(pos);
+          String idString = String.valueOf(field.fieldId());
+          if (defaultValue.has(idString)) {
+            defaultRecord.set(pos, fromJson(field.type(), defaultValue.get(idString)));
+          }
+        }
+        return defaultRecord;
+      default:
+        throw new IllegalArgumentException(String.format("Type: %s is not supported", type));
+    }
+  }
+
+  public static Object fromJson(Type type, String defaultValue) {
+    try {
+      JsonNode defaultValueJN = JsonUtil.mapper().readTree(defaultValue);
+      return fromJson(type, defaultValueJN);
+    } catch (IOException e) {
+      throw new IllegalArgumentException("Failed to parse: " + defaultValue + "; reason: " + e.getMessage(), e);
+    }
+  }
+
+  public static String toJson(Type type, Object defaultValue) {
+    return toJson(type, defaultValue, false);
+  }
+
+  public static String toJson(Type type, Object defaultValue, boolean pretty) {
+    try {
+      StringWriter writer = new StringWriter();
+      JsonGenerator generator = JsonUtil.factory().createGenerator(writer);
+      if (pretty) {
+        generator.useDefaultPrettyPrinter();
+      }
+      toJson(type, defaultValue, generator);
+      generator.flush();
+      return writer.toString();
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
+  public static void toJson(Type type, Object defaultValue, JsonGenerator generator)
+      throws IOException {
+    if (defaultValue == null) {
+      generator.writeNull();
+      return;
+    }
+    switch (type.typeId()) {
+      case BOOLEAN:
+        Preconditions.checkArgument(
+            defaultValue instanceof Boolean, "Invalid default %s value: %s", type, defaultValue);
+        generator.writeBoolean((Boolean) defaultValue);
+        break;
+      case INTEGER:
+        Preconditions.checkArgument(
+            defaultValue instanceof Integer, "Invalid default %s value: %s", type, defaultValue);
+        generator.writeNumber((Integer) defaultValue);
+        break;
+      case LONG:
+        Preconditions.checkArgument(defaultValue instanceof Long, "Invalid default %s value: %s", type, defaultValue);
+        generator.writeNumber((Long) defaultValue);
+        break;
+      case FLOAT:
+        Preconditions.checkArgument(defaultValue instanceof Float, "Invalid default %s value: %s", type, defaultValue);
+        generator.writeNumber((Float) defaultValue);
+        break;
+      case DOUBLE:
+        Preconditions.checkArgument(defaultValue instanceof Double, "Invalid default %s value: %s", type, defaultValue);
+        generator.writeNumber((Double) defaultValue);
+        break;
+      case DATE:
+        Preconditions.checkArgument(
+            defaultValue instanceof Integer && (Integer) defaultValue >= 0,
+            "Invalid default %s value: %s",
+            type,
+            defaultValue);
```

Comment:
Style (minor): the format arguments should be on the same line as the format string unless they need to be wrapped, not one per line.

Reply:
fixed